src/signal_processing_algorithms/energy_statistics/energy_statistics.py

```diff
@@ -0,0 +1,366 @@
+"""Energy statistics."""
+from dataclasses import dataclass
+from typing import List, Union, Optional
+
+import numpy as np
+from more_itertools import pairwise
+from signal_processing_algorithms.energy_statistics.cext_calculator import C_EXTENSION_LOADED, calculate_diffs, \
+    calculate_t_values
+
+
+def _get_distance_matrix(series: np.ndarray, use_c_if_possible=True) -> np.ndarray:
+    """
+    Return the matrix of pairwise distances between x and y.
+
+    :param x: An m x n array of m observations for n variables.
+    :param y: An l x n array of l observations for n variables.
+    :return: An m x l array where (i,j)th value is the distance between the observation
+    at i-th row of x and j-th row of y.
+    """
+    if use_c_if_possible and C_EXTENSION_LOADED and series.shape[1] == 1:
+        return calculate_diffs(series.flatten())
+    else:
+        return np.linalg.norm(series[:, np.newaxis] - series, axis=2)
+
+
+@dataclass
+class EnergyStatistics:
+    """
+    Class for representing Energy Statistics.
+
+    e - E-statistic
+    t - Test statistic
+    h - E-coefficient of inhomogeneity
+    """
+
+    e: float
+    t: float
+    h: float
+
+
+@dataclass
+class EnergyStatisticsWithProbabilities(EnergyStatistics):
+    """Class for representing Energy Statistics and permutation test result."""
+
+    e_pvalue: float
+    t_pvalue: float
+    h_pvalue: float
+
+
+def _calculate_stats(x: float, y: float, xy: float, n: int, m: int) -> (float, float, float):
+    """
+    Calculate the q value from the terms and coefficients.
+
+    :param xy: The sum of differences between sample distributions X, Y.
+    :param x: The sum of differences within sample distribution X.
+    :param y: The sum of differences within sample distribution Y.
+    :param n: The length of sample distribution X.
+    :param m: The length of sample distribution Y.
+
+    :return: The q value generated from the terms.
+    """
+
+    xy_avg = xy / (n*m) if n > 0 and m > 0 else 0
+    x_avg = x / (n**2) if n > 0 else 0
+    y_avg = y / (m**2) if m > 0 else 0
+
+    # E-statistic
+    e = 2 * xy_avg - x_avg - y_avg
+    # Test statistic
+    t = (n * m / (n + m)) * e
+    # E-coefficient of inhomogeneity
+    h = (e / (2 * xy_avg)) if xy_avg > 0 else 0.0
+    return e, t, h
+
+
+def _calculate_t_stats(diffs: np.ndarray, use_c_if_possible=True) -> np.ndarray:
+    """
+    Find e-statistics values given a difference matrix.
+
+    :param diffs: The difference matrix.
+    :return: The qhat values.
+    """
+    if use_c_if_possible and C_EXTENSION_LOADED:
+        return calculate_t_values(diffs)
+
+    statistics = np.zeros(len(diffs), dtype=np.float64)
+
+    # We will partition our signal into:
+    # X = {Xi; 0 <= i < tau}
+    # Y = {Yj; tau <= j < len(signal) }
+    # and look for argmax(tau)Q(tau)
+
+    # sum |Xi - Yj| for i < tau <= j
+    xy = 0
+    # sum |Xi - Xj| for i < j < tau
+    x = 0
+    # sum |Yi - Yj| for tau <= i < j
+    y = np.sum(diffs)
```

Comment:
Technically we could make this still do just half of the diff matrix and then double the terms due to symmetry arguments. We can probably leave that for a future optimization ticket.

Reply:
Made the optimization